src/backend/utils/adt/network.c

```diff
@@ -389,6 +416,415 @@ network_cmp(PG_FUNCTION_ARGS)
+	}
+
+	/*
+	 * For IPv6, create an integer from either the first 4 or 8 bytes of the
+	 * IP (depending on whether we're on 32 or 64-bit architecture). For IPv4,
+	 * always use all 4 bytes.
+	 *
+	 * We're consuming an array of char, so make sure to byteswap on little
+	 * endian systems (an inet's IP array emulates big endian in that the
+	 * first byte is always the most significant).
+	 */
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		ipaddr_int = *((Datum *) ip_addr(authoritative));
+		ipaddr_int = DatumBigEndianToNative(ipaddr_int);
+	}
+	else
+	{
+		uint32		ipaddr_int32 = *((uint32 *) ip_addr(authoritative));
+#ifndef WORDS_BIGENDIAN
+		ipaddr_int = pg_bswap32(ipaddr_int32);
+#endif
+	}
+
+	netmask_int = ipaddr_int;
+	subnet_int = (Datum) 0;
+	datum_subnet_size = 0;
+
+	/* Extract subnet bits from ipaddr_ints if there are some present. */
+	datum_size_left = SIZEOF_DATUM * BITS_PER_BYTE - ip_bits(authoritative);
+	if (datum_size_left > 0)
+	{
+		/*
+		 * Number of bits in subnet. e.g. An IPv4 that's /24 is 32 - 24 = 8.
+		 *
+		 * However, only some of the bits may have made it into the fixed
+		 * sized datum, so take the smallest number between bits in the subnet
+		 * and bits left in the datum.
+		 */
+		datum_subnet_size = Min(ip_maxbits(authoritative) - ip_bits(authoritative),
+								datum_size_left);
+
+		if (datum_subnet_size > 0)
+		{
+			/*
+			 * This shift creates a power of two like `0010 0000`, and
+			 * subtracts one to create a bitmask for an IP's subnet bits like
+			 * `0001 1111`.
+			 */
+			Assert(datum_subnet_size < SIZEOF_DATUM * BITS_PER_BYTE);
+			Datum		subnet_bitmask = (((Datum) 1) << datum_subnet_size) - 1;
+
+			/*
+			 * AND the bitmask with the IP address' integer to truncate it
+			 * down to just the bits after the netmask.
+			 */
+			subnet_int = ipaddr_int & subnet_bitmask;
+
+			/*
+			 * Integer representation for the netmask is the IP's integer
+			 * minus whatever we calculated for the subnet.
+			 */
+			netmask_int = ipaddr_int - subnet_int;
+		}
+	}
+
+#if SIZEOF_DATUM == 8
+
+	if (ip_family(authoritative) == PGSQL_AF_INET6)
+	{
+		/*
+		 * IPv6 on 64-bit architecture: keep the most significant 63 netmasked
+		 * bits.
+		 */
+		res |= netmask_int >> ABBREV_BITS_INET_FAMILY;
+	}
+	else
+	{
+		Datum netmask_shifted;
+
+		/*
+		 * IPv4 on 64-bit architecture: keep all 32 netmasked bits, netmask
+		 * size, and most significant 25 subnet bits (see diagram above for
+		 * more detail).
+		 */
+		Datum		netmask_size_and_subnet = (Datum) 0;
+
+		/*
+		 * an IPv4 netmask has a maximum value of 32 which takes 6 bits to
+		 * contain
+		 */
+		Datum		netmask_size = (Datum) ip_bits(authoritative);
+
+		Assert(netmask_size <= ip_maxbits(authoritative));
+		netmask_size_and_subnet |= netmask_size << ABBREV_BITS_INET4_SUBNET;
+
+		/*
+		 * if we have more than 25 subnet bits of information, shift it down
+		 * to the available size
+		 */
+		if (datum_subnet_size > ABBREV_BITS_INET4_SUBNET)
```

Comment:
Do you need this branch?

Reply:
I could replace with something like:

``` c
subnet_int >>= Max(0, datum_subnet_size - ABBREV_BITS_INET4_SUBNET);
```

But I'd say that's quite a bit less clear than just having the conditional. Thoughts?

Reply:
Okay nevermind — I changed it in the end and clarified about the <= 0 condition via a comment. Thanks!